contracts/schemes/GenericSchemeMultiCall.sol

```diff
@@ -0,0 +1,167 @@
+pragma solidity 0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "@daostack/infra/contracts/votingMachines/IntVoteInterface.sol";
+import "@daostack/infra/contracts/votingMachines/ProposalExecuteInterface.sol";
+import "../votingMachines/VotingMachineCallbacks.sol";
+
+/**
+ * @title GenericScheme.
+ * @dev  A scheme for proposing and executing calls to multiple arbitrary function
+ * on whitelisted contract on behalf of the organization avatar.
+ */
+contract GenericSchemeMulticall is VotingMachineCallbacks, ProposalExecuteInterface {
+    event NewMultiCallProposal(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        bytes[]   _callData,
+        uint[] _value,
+        string  _descriptionHash,
+        address[] _contractToCall
+    );
+
+    event ProposalExecuted(
+        address indexed _avatar,
+        bytes32 indexed _proposalId
+    );
+
+    event ProposalCallExecuted(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        address _contractToCall,
+        bytes _callDataReturnValue
+    );
+
+    event ProposalExecutedByVotingMachine(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        int256 _param
+    );
+
+    event ProposalDeleted(address indexed _avatar, bytes32 indexed _proposalId);
+
+    // Details of a voting proposal:
+    struct MultiCallProposal {
+        address[] contractToCall;
+        bytes[] callData;
+        uint256[] value;
+        bool exist;
+        bool passed;
+    }
+
+    mapping(bytes32=>MultiCallProposal) public organizationProposals;
+
+    IntVoteInterface public votingMachine;
+    bytes32 public voteParams;
+    Avatar public avatar;
+
+    /**
+     * @dev initialize
+     * @param _avatar the avatar to mint reputation from
+     * @param _votingMachine the voting machines address to
+     * @param _voteParams voting machine parameters.
+     */
+    function initialize(
+        Avatar _avatar,
+        IntVoteInterface _votingMachine,
+        bytes32 _voteParams
```

Comment:
how about adding here a white listed targets contracts? for security reasons.

Reply:
@orenyodfat as we would need to call allowance() on some tokens we don't know at initialization we won't be able to whitelist specific contract addresses. Thinking about a possible solution for restriction right now.

Reply:
losing a bit from generality , one option is to add explicit none restricted `approval` token calls as part of a proposal .
using https://github.com/daostack/arc/blob/master/contracts/controller/Controller.sol#L460

Reply:
We could still keep the generality if we would include an encoded call within the proposalCalls function that would execute externalTokenApproval() on the controller before sending funds. In that way we could still use a contract whitelist as we would always use the controller (known address) for approvals – correct?

Reply:
> @orenyodfat we could still keep the generality if we would include an encoded call within the proposalCalls function that would execute externalTokenApproval() on the controller before sending funds. In that way we could still use a contract whitelist as we would always use the controller (known address) for approvals – correct?

no. because the avatar is not allowed to call controller . it is not registered scheme

Reply:
> @orenyodfat we could still keep the generality if we would include an encoded call within the proposalCalls function that would execute externalTokenApproval() on the controller before sending funds. In that way we could still use a contract whitelist as we would always use the controller (known address) for approvals – correct?

no . because the avatar is not allowed to call controller .(it is not a register scheme)

Reply:
@orenyodfat the avatar wont call the controller it would called an ERC20 contract and execute the approve function.

Reply:
But taht will require the ERC20 token to be listed in the whitelisted contracts, right? this way the controller can call the approve function in any token contract, right? or am I missing something?

Reply:
@nicoelzer  does any how the plan is to allow the avatar to transfer any erc20 token its own?. this might be the same as approval ... if this is the case we might do not need this special approval staff.
If the case is that only approval is needed for any erc20 so we can avoid white listing the erc20 and have the solution as it is now .